aggregate_spatial.json

```
@@ -0,0 +1,130 @@
+{
+    "id": "aggregate_spatial",
+    "summary": "Compute zonal statistics for geometries",
+    "description": "Aggregates zonal statistics for one or more geometries (e.g. points or polygons) over the spatial dimensions.\n\nThe process considers all pixels for which the point at the pixel center intersects with the corresponding geometry (as defined in the Simple Features standard by the OGC).\n\nThe data cube must have been reduced to only contain two raster dimensions and a third dimension the values are aggregated for, for example the temporal dimension to get a time series. Otherwise this process fails with the `TooManyDimensions` error.\n\nThe number of total and valid pixels is returned together with the calculated values.",
```

```suggestion
    "description": "Aggregates zonal statistics for one or more geometries (e.g. points or polygons) over the spatial dimensions.\n\nThe process considers all pixels which touch the corresponding geometry (as defined in the Simple Features standard by the OGC).\n\nThe data cube must have been reduced to only contain two raster dimensions and a third dimension the values are aggregated for, for example the temporal dimension to get a time series. Otherwise this process fails with the `TooManyDimensions` error.\n\nThe number of total and valid pixels is returned together with the calculated values.",
```

This would mean that in some cases pixels are considered multiple times in aggregations, e.g. counting all pixels in Germany and The Netherlands would in most cases count all pixels on the border for both Germany and The Netherlands.

---

FYI: At VITO we have a use case where the possibility of pixels being in multiple aggregation buckets simultaneously result in important performance/implementation  issues.

---

See https://open-eo.github.io/openeo-api/glossary/#aggregate-reducing-resolution - it says the assignment to groups is **unique**.

As a side note - I suggest we stop [treating pixels as little squares](https://open-eo.github.io/openeo-api/glossary/#aggregate-reducing-resolution) and just go with pixel centres whenever we need to think about the pixel in continuous space.

---

Right. So how would that work with lines/points? It seems we need to define separate behavior for polygons (pixel centers) and lines/points (pixel centers in a buffer around the feature? Are assignment in this case always unique?). So I'm questioning the original proposal to merge all feature types into a single process...

---

With lines and points, it is quite difficult to enforce unique assignment unless we somehow prevent "occupied" pixels from being used. If we use center-inside rule for polygons, a nearest-pixel rule could be seen as analogous, e.g. for point set `A`:
- find the pixel centre `Pxy` nearest to point `Ai` (this is equivalent to finding the square cell that `Ai` falls into - squares are the [voronoi cells](https://en.wikipedia.org/wiki/Voronoi_diagram) of the centre-point grid)
- check if `Ai` is the nearest point from set `A` to pixel centre `Pxy` (to ensure uniqueness)
- if `Ai` is not the nearest, then `Ai` gets no pixels

Another option would be to include a parameter `radius` (with default value `sqrt(2) * gsd` where `gsd` is raster ground sampling distance) and include, for each point, all pixels `Pxy` within `radius` from point `Ai`, for which `Ai` is the nearest point from set `A`. (this second check is not necessary if the set `A` does not contain pairs of points that are less than `2 * radius` apart).

For linestrings, we can do similar - either use all pixels whose cells intersect the line `Ti` and for which the line `Ti` is the nearest of all the lines in set `T` (i.e. select all pixels whose centres are nearest to at least one point on the linestring); or use a radius (default could again be `sqrt(2) * gsd`) with the same uniqueness check (which is not necessary if any two lines from `T` are at least `2 * radius` apart).

While these process would fit our description for the *aggregate* process, I'm not sure whether aggregation is really what the user wants when calling this with a non-polygon geometry. If I understand the raster as a discrete sampling of some underlying smooth function, then:
- for points, the more common operation one would expect would be that of `resample` (i.e. I would like to get a sampling of that underlying function at locations that are different from the ones provided in the raster); the value will usually be better if created from raster interpolation rather than just selection of a single raster value;
- for lines, I imagine the need of the user is to get summary statistics of the values that the *underlying smooth function* assumes along the provided linear feature, which might be again better served by using an interpolation of the raster (and then some aggregation for statistics) rather than all-or-nothing aggregation of some neighbouring pixels

I would support removing the point and linestring features from the `aggregate_spatial` process and creating a more targeted process for the point-sampling and summary-along-a-linestring use-cases.

---

I agree. I think polygon and line/point are too different and should be separate processes.